### ISSUE TITLE
Partially broken css/js

### ISSUE THREAD

**Issue by Sashkan** [nathanvda/cocoon#396]
Hi everyone !

I have some troubles understanding the bahaviour of newly created inputs, on both CSS and jQuery.

My form works just fine, I can add and remove associations, but:
- first of all, the CSS seems to be partially broken. If my partial looks like this:
``` erb
<tr class="expense">
  <td><%= f.input :title, label: false, required: true, placeholder: t('simple_form.placeholders.expense.title') %></td>
  <td><%= f.input :day_based, input_html: { class: 'quote-checkbox-day-based', placeholder: t('simple_form.placeholders.expense.day_based') } %></td>
  <td><%= f.input :days, label: false, required: true, input_html: { class: 'quote-expense-days', placeholder: t('simple_form.placeholders.expense.days') } %></td>
  <td>

  </td>
  <td><%= f.input :price, label: false, required: true, input_html: { class: 'quote-expense-price expense-price-input', placeholder: t('simple_form.placeholders.expense.price') } %></td>
  <td>
    <%= link_to_remove_association(t('simple_form.buttons.remove_expense'), f,
    { wrapper_class: 'expense', class: 'btn btn-danger btn-block' }) %>
    </td>
</tr>

``` 
The newly rendered partial will have all the matching classes and CSS, but the table won't (see picture)

![screenshot from 2016-11-08 11-28-25](https://cloud.githubusercontent.com/assets/6276267/20095616/7e34d822-a5a6-11e6-80b4-b8e795029e12.png)

- Second thing is (but I think it's linked to the first one), I cannot get to select the newly created inputs as dom elements. So, if I use this little pieceo f code:

``` js
  $('.expense-price-input').keyup(function(){
    console.log('hello');
  });
```
On keyup for the existing inputs, I will have a "hello" in my console logs, but this won't happen for the newly added partials. 

How comes so ?

**Comment by nathanvda** [comment]
Of course, from a screenshot I cannot deduce 1) how your html looks, and 2) how you want it to look.  
I do not even know which part of the screenshot is added dynamically, but imho this can only be explained by 1) your html/view does not use the correct classes, or 2) your css does not do what you think it should do. Depending on the framework (if any) you are using, it is even possible some style is applied using js? (which would be unusual, but it is possible). 

Note: you refer to an unstyled "table" and you only show a single row (`tr`) ? Things to do: check the source html, because dynamic insertion with cocoon will not add/remove any classes. Another option is that you are not rendering the same partial (also not shown, so again guess-work for me here). 


Regarding your javascript-behaviour. The `$(.expense-price-input')` selector is only run once, at the start of the page, so newly/dynamically added items will never be in that set (so that will never work). 

This is however pretty basic jquery feature, a simple way to solve this is: 

    $('form').on('keyup', '.expense-price-input', function() { ....})

To conclude you will be better off asking such questions, regarding basic html/css/jquery on stackoverflow.com as you will get help quicker.

**Comment by Sashkan** [comment]
Yes, my bad, I'm aware I've been a bit greedy on explanations.
I'm using rails 5, and here are my partials

First one is the ```_expense_fields.html.erb``` partial
``` erb
<tr class="expense">
  <td><%= f.input :title, label: false, required: true, placeholder: t('simple_form.placeholders.expense.title') %></td>
  <td><%= f.input :day_based, input_html: { class: 'quote-checkbox-day-based', placeholder: t('simple_form.placeholders.expense.day_based') } %></td>
  <td><%= f.input :days, label: false, required: true, input_html: { class: 'quote-expense-days', placeholder: t('simple_form.placeholders.expense.days') } %></td>
  <td><%= f.input :price, label: false, required: true, input_html: { class: 'quote-expense-price expense-price-input', placeholder: t('simple_form.placeholders.expense.price') } %></td>
  <td>
    <%= link_to_remove_association(t('simple_form.buttons.remove_expense'), f,
    { wrapper_class: 'expense', class: 'btn btn-danger btn-block' }) %>
    </td>
</tr>

```
which is rendered in the ```expense_category_fields.html.erb``` one

```erb
<tbody class="category">
  <tr>
    <th colspan="4" class="text-right"><%= f.input :title, label: false, class: 'text-right' %></th>
    <th>
    <%= link_to_remove_association(t('simple_form.buttons.remove_expense_category'), f,
  { wrapper_class: 'category', class: 'btn btn-danger btn-block' }) %>
    </th>
  </tr>
  <div id='expenses'>
  <%= f.simple_fields_for :expenses do |expense| %>
    <div>
      <%= render 'briefs/expense_fields', f: expense %>
    </div>
  <% end %>
  <tr>
    <th colspan="6">
      <%= link_to_add_association t('simple_form.buttons.add_expense'), f, :expenses,
      partial: 'briefs/expense_fields', render_options: { wrapper: 'expense' }, class: 'add-expense-link btn btn-success btn-block' %>
    </th>
  </tr>
  </div>
</tbody>

```

Now, from what I read in the issues and setup guide, if I use bootstrap and simple-form, I have to make sure to create wrappers, which I did in my ```simple_form_bootstrap.rb``` file:

``` rb

  config.wrappers :expense, :class => 'expense', :error_class => :error do |b|
    b.use :placeholder
    b.use :html5
    b.use :input, class: 'form-control', :wrap_with => { :tag => 'td' }
    b.use :label
  end
```
I thought that this would be enough to make sure that the newly added expense fields would look just like the previous ones, especially by using the ```:wrap_with => { :tag => 'td' }``` (On the attached picture, last line is the only freshly added line, the one called "Expense 4", but as you can see, it doesn't behave as the other table rows)

**Comment by nathanvda** [comment]
I am not sure why you present this code to me. It looks good. As you are probably aware (because I mentioned what could be wrong and it obviously it is not). Did you check the html (as i suggested) ? I am guessing your row is not inserted where you think it should. Your html does not like the canonical example (the default location of `link_to_add_association` is correct if you use the html structure as shown  in the examples).

Where your partial is inserted is perfectly configurable using the `data-association-insertion-node` (which can be a function for maximum flexibility). Check the [documentation](https://github.com/nathanvda/cocoon#link_to_add_association)

**Comment by Sashkan** [comment]
Hi ! Thanks for your answer, it came from the HTML (At least, the CSS part), but I'm still struggling with the JS part.

I'm using cocoon to create quotes for my users. A ```Quote``` has_many ```Expense Categories```, and each ```Expense Category``` has many ```Expenses```

I added some JS, so that everytime my user changes the value of an expense, it updates the total price in another input, but this doesn't work with the newly created inputs.

I made sure these inputs all had the same classes (.expense .price-input), and then tried with this little piece of code:
``` js
  $('.expense .price-input').keyup(function () {
    alert('OK');
  });
```
When I type something in the existing inputs, OK is displayed, but when I create an expense using the ```link_to_add_association```, typing something in the newly created input won't trigger the alert. Is there something to initialize for the JS ?
